### PR TITLE
BugFix: Util.parseLong

### PR DESCRIPTION
**Pull request by htmldoug**
Currently:
```scala
upickle.default.read[Long](""""a"""") ==> 49L
```

Port of https://github.com/rallyhealth/weePickle/pull/88.

### CODE REVIEW

**Comment by htmldoug** [comment]
That first commit should have failed in CI. Seems like tests aren't being run. Maybe related to #359.

**Review by lihaoyi — APPROVED** [review]

**Comment by lolgab** [comment]
Rebased this PR after merging https://github.com/com-lihaoyi/upickle/pull/359 which fixes the CI.

**Comment by lolgab** [comment]
@htmldoug Can you check the failing test?

**Comment by htmldoug** [comment]
Thanks so much for fixing up the build and tests. Good thing too, since my assumptions about this method were wrong.

Should be good now, although I've jumped around with approaches and pickle flavors so much that I'd appreciate another set of eyes on it before merging.

File: core/src/upickle/core/Util.scala
Comment (lolgab) [inline core/src/upickle/core/Util.scala:75]: Sorry, I can't understand this check here.
You make the or of the `start`, `end` etc. and check if the result is `< 0`. Why?
Also are you sure there isn't redundancy between this check and  the one in line 83?

Comment (lolgab) [reply core/src/upickle/core/Util.scala:75]: I think I got the meaning. You do the OR of all these numbers since none of them should have the sign bit at `1`. So this is equivalent to:
```scala
if (start < 0 || end < 0  || (end - start) < 0 || (cs.length - end) < 0) throw new IndexOutOfBoundsException
```
About the redundancy of the checks, the one in line 83 (`size <= 0`) checks if you're trying to parse just a `"-"` which should fail. Right?

**Comment by lolgab** [comment]
I'm approving and merging the PR as it is.
Any improvements can be done later as a follow-up PR.

File: core/src/upickle/core/Util.scala
Comment (htmldoug) [reply core/src/upickle/core/Util.scala:75]: Right. The `size <= 0` check should result in an exception thrown for `"-"`.

Credit to [java.io.FilterOutputStream](https://github.com/openjdk/jdk/blob/cec6c068b03d890312e50b448fbc26102c635249/src/java.base/share/classes/java/io/FilterOutputStream.java#L133) for the arcane bitwise bounds check. I assume it was written that way for performance, but I haven't measured if the difference is actually noticeable over your equivalent and clearer version. That class says `@since JDK1.0` which was 1996. CPU branch prediction and the JVM implementation have improved since then.

The bounds check isn't really related to the fix, but since I found one bug and didn't have a good understanding of how this is used, I wanted to nail down the assumptions around inputs. I think `parseLong` is used deep in the parsing code (not public API) so the check could probably be removed if needed, although I didn't expect the cost of keeping it to be significant.

**Comment by htmldoug** [comment]
Thanks for the review and merge, @lolgab!

**Pull request merged**
